### PR TITLE
Version 0.9

### LINKED ISSUE #235: Docs: only reward(m, s, a) shows up
Only the docs for reward(m, s, a) show up on the API page. The docs for reward(m, s, a, sp) do not show up.

### LINKED ISSUE #237: Consider moving `initialstate_distribution` to the common section in the documentation
When implementing the `POMDP` interface -- even when implementing the generative version -- it seems to me that there is barely anything that one can do with this model unless `initialstate_distribution` is implemented. E.g. when running a `HistoryRecorder`, even when using the `NothingUpdater` the code fails, since the initial belief distribution can not be determined (trying to call `initialstate_distribution`). So the common workaround seems to be to implement some custom distribution type and then define `Base.rand` on this type which simply calls `initialstate` from the generative interface.

Maybe `initiastate_distribution` should be moved to the "common" section of the POMDPs.jl API documentation? Or maybe one could place a hint somewhere that the above workaround exists. @zsunberg suggested that this workaround could be placed in [`POMDPModelTools.jl`](https://github.com/JuliaPOMDP/POMDPModelTools.jl) in a generalized form.

Feel free to come up with a better title for this issue.

### LINKED ISSUE #263: The future of requirements and `implemented`
In version 0.8, `implemented` is slightly broken. In some cases (notably `gen(::DBNOut, ...)`, instead of returning `true` or `false`, it returns `missing`, which will show up as a `[?]` in checked lists. This is because it was too difficult to determine if the appropriate methods are implemented because it is difficult to determine the types of samples from arbitrary DBN Nodes.

The whole requirements system (of which `implemented` is a part) was introduced to help new users with the steep learning curve of POMDPs.jl. As QuickPOMDPs becomes more capable, I think this will start to be less important. Moreover, the requirements system and `implemented` contain information redundant with the source code, which makes it a major headache to maintain. 

Since `implemented` is now less complete, and I don't think we should put in the effort to make it perfect (that would mean analyzing a bunch of intermediate forms of the code), I propose that (in the medium term) we should move the requirements functionality out of the core package into POMDPTesting or a linter package.

### LINKED ISSUE #280: Update documentation for defining a solver
The "defining a solver" documentation is out of date. It uses `type` for instance.

### LINKED ISSUE #306: Remove customizable DDN in POMDPs.jl v0.9
Yesterday, I did a lot of thinking about DynamicDecisionNetworks.jl, and I concluded that it needs to be designed from the ground up independently of POMDPs.jl. Julia has changed significantly since we originally designed POMDPs.jl, and we have learned many lessons, so I think we can make a much better interface if we start over. I am confident that these two things will be interoperable (i.e. it will be easy to create a DDN from a POMDP and it will be easy to turn a DDN into a POMDP), but we should not try too hard to stick to POMDPs.jl design principles (as @rejuvyesh suggested, we should take inspiration from other packages like Turing.jl or ForneyLab.jl).

**As a consequence of this, I propose that we completely remove the DDN customization functionality from POMDPs.jl 0.9 (#297) and implement it later** (instead of making POMDPs.jl 0.9 depend on DynamicDecisionNetworks.jl as proposed in #292).

Conceptually, all POMDPs will have the same standard DDN, and they will all have an `:info` node for outputting debug information. We will get rid of `gen(::DDNNode, ...)` (since I don't think anyone used it). This will leave the `gen` function that returns a `NamedTuple`, and the explicit `transition`, `observation`, and `reward` functions for defining problems (if people want to have a generative model for the observation, but an explicit transition model, they can use a lazy distribution #269). The `@gen` macro will still be used for calling generative models.

I think this will result in a nice stable POMDP and MDP interface that we can support forever, and we can explore new ideas in DynamicDecisionNetworks.jl.

Does anyone have any objections to this?

### LINKED ISSUE #308: Consistent distribution semantics (change initialstate? change action?)
I just [added `ImplicitDistribution` to POMDPModelTools](https://github.com/JuliaPOMDP/POMDPModelTools.jl/pull/37/). This makes it easier to create a distribution object when you can't write down the explicit distribution, but only have a function to sample from the distribution.

We may want to change a couple things in light of this:

## 1. `initialstate`

This makes `initialstate(m, rng)` somewhat unnecessary because now it is easy enough to do
```julia
initialstate_distribution(m::MyPOMDP) = ImplicitDistribution(rng -> #=whatever used to be in initialstate=#)
```

I see two options for eliminating one of these redundant functions:
1. Make `initialstate` return a distribution
    - Pros: Consistent with `transition` and `observation`
    - Cons:
        - Inconsistent with previous versions (but luckily there is a clear deprecation pattern: `@deprecate initialstate(m, rng) rand(rng, initialstate(m))` and `@deprecate initialstate_distribution(m) initialstate(m)`)
        - Breaks `QuickMDP(initialstate=1, ...)`
2. Get rid of `initalstate` altogether (keep `initialstate_distribution`)
    - Pros: No change to meaning of `initialstate`
    - Cons: Inconsistent with `transition` and `observation`
3. Get rid of `initialstate` and `initialstate_distribution` and create a new function `initial(m::Union{MDP, POMDP})` that returns the initial state distribution.
    - Pros: Easy deprecation pattern and no confusion
    - Cons: Name is a bit less clear
    - Notes: Still need to figure out what to do about `initialobs`

## 2. `action`
If we are going to make everything distribution-focused, should we also make `action` return a distribution?

Options:
1. Make no changes
    - Pros: No changes
    - Cons:
        - Inconsistent with `transition` and `observation` (but this does not seem too bad since policies are different from POMDPs)
        - No standard interface for stochastic policies.
2. Make `action` return a distribution
    - Pros: Consistent with `transition` and `observation`
    - Cons:
        - Big change with no clear deprecation pattern
        - `rand(action(policy, b))` does not feel that clean
        - Breaks a bunch of things like `FunctionPolicy`
2. Add `action_distribution(policy, b)` or just `distribution(policy, b)`
    - Pros: Standard interface for stochastic policies
    - Cons: Inconsistent with `transition` and `observation`
    - Notes: Some details would need to be ironed out, e.g. should we have `action(policy, b, rng)`? What is the default fallback pattern?

----

Though I'm very undecided at this point, my initial feeling is that we should introduce `initial` (option (3)) and not make any changes to `action` (option (1)).

### LINKED ISSUE #309: Role of `gen` and `@gen`
Discussion in #308 has raised the issue of the role of `gen`.

I see the following requirements ([Req] denotes an absolute requirement, [Nice-to-have] is not necessarily needed):

## Requirements
1. [Req] A way to handle the case where `sp`, `o` and `r` are generated at the same time
2. [Req] An explicit interface
3. [Req] A way to specify a generative transition model and explicit observation model (for POMCPOW)
4. [Req] A way to specify a generative observation model and explicit transition model (for QMDP with a complex observation model)
5. [Nice-to-have] Concise syntax for specifying a completely generative model
6. [Nice-to-have] Concise syntax for calling the generative model

In my mind, `gen` satisfies 1 and 5, `transition`, `observation`, and `reward` (TOR) with optional implicit distributions satisfies 2, 3, and 4. `@gen` satisfies 6.

Previously, I thought we needed `gen(DDNNode, ...)` to handle 3 and 4, but now I think `ImplicitDistribution` is better.

The documentation could emphasize the "generative interface" less. Instructions for what model to implement could follow:
- If an explicit transition **or** observation model is available, implement TOR, possibly using `ImplicitDistribution` if one of the components is generative.
- If the entire model is generative, implement `gen`.

`gen` (or at least a function that returns `(sp= , r= )`) is a very nice syntax for QuickPOMDPs, expecially if people just want to use MCTS, so I think we should keep it.

I'm very open to changing this if we have a good reason and can support all of the requirements.

### PR DESCRIPTION
**Pull request by zsunberg**
This branch will currently
- fix #306 
- fix #308 
- fix #309 
- fix #263 
- fix #237 
- fix #235
- fix #280

### CODE REVIEW

**Comment by zsunberg** [comment]
To keep track of fixes that this branch currently has, it will currently
- fix #306 
- fix #308 
- fix #309 
- fix #263 
- fix #237 
- fix #235

**Comment by zsunberg** [comment]
Ok, I think this is ready to merge and release. The solvers and models will have to be updated afterwards, but I think this will be nearly all find-and-replace (see #312 and #313). I have already updated some of the packages and it has been straightforward, and I think nearly all models and solvers compatible with POMDPs 0.8 will run in POMDPs 0.9 with deprecation warnings.

I know people are busy right now, so not expecting a ton of feedback, but any is welcome! I have requested "reviews" from @MaximeBouton, @rejuvyesh, and @lassepe , but I am really just looking for any broad questions and concerns before merging this - you only need to do a detailed low level code review if you really want to :)

I'll wait until next week for any comments and then merge and register this and start on package upgrades.

**Comment by lassepe** [comment]
You may want so add your "fix list" above to the pr description so that these issues get closed automatically on merge. I think it does not work of you mention them in a comment.

File: test/runtests.jl [hunk 1]
```diff
@@ -94,3 +45,34 @@ end
 end
 
 POMDPs.add_registry()
```
Comment (rejuvyesh) [inline test/runtests.jl:47]: Deprecated?

File: docs/src/def_pomdp.md [hunk 2]
```diff
@@ -7,7 +7,7 @@ Since POMDPs.jl was designed with performance and flexibility as first prioritie
...
 - [POMDPExamples.jl](https://github.com/JuliaPOMDP/POMDPExamples.jl) provides tutorials for defining problems. 
 - [The Tabular(PO)MDP model](https://github.com/JuliaPOMDP/POMDPExamples.jl/blob/master/notebooks/Defining-a-tabular-POMDP.ipynb) from [POMDPModels.jl](https://github.com/JuliaPOMDP/POMDPModels.jl) allows users to define POMDPs with matrices for the transitions, observations and rewards.
-- The [`gen` function](@ref generative_doc) is the easiest way to wrap a pre-existing simulator from another project or written in another programming language so that it can be used with POMDPs.jl solvers and simulators. See also [RLInterface.jl](https://github.com/JuliaPOMDP/RLInterface.jl) for an even higher level interface for simulators where the state is not accessible.
+- The [`gen`](@ref) function is the easiest way to wrap a pre-existing simulator from another project or written in another programming language so that it can be used with POMDPs.jl solvers and simulators. See also [RLInterface.jl](https://github.com/JuliaPOMDP/RLInterface.jl) for an even higher level interface for simulators where the state is not accessible.
```
Comment (MaximeBouton) [inline docs/src/def_pomdp.md:10]: Do we really want to advertise RLInterface here? I think people should implement `gen` before RLInterface. The only solver using RLInterface is DeepQLearning.jl for now and it is still relying on functions that are not part of `RLInterface.jl`.

File: README.md [hunk 3]
```diff
@@ -11,13 +11,19 @@
...
+There are [several ways to define and interact with (PO)MDPs](http://juliapomdp.github.io/POMDPs.jl/stable/def_pomdp): transition and observation distributions and rewards can be defined with explicit probabiliyt distributions or implicitly with a function that samples from the distribution, or all of the dynamics can be defined in a single step simulator function: (s', o, r) = G(s,a). Problems may also be defined with probability [tables](http://juliapomdp.github.io/POMDPs.jl/stable/explicit/#tabular-form-explicit-pomdp), and the [QuickPOMDPs interfaces](https://github.com/JuliaPOMDP/QuickPOMDPs.jl) make defining simple problems easier.
```
[the remark this answers is not in the document]
Comment (zsunberg) [reply README.md]: I haven't seen any way to do this in the README. One of the links was wrong anyways though.

File: test/runtests.jl (hunk 1 above)
Comment (zsunberg) [reply test/runtests.jl]: Yes, thanks

File: docs/src/def_pomdp.md (hunk 2 above)
Comment (zsunberg) [reply docs/src/def_pomdp.md:10]: Yeah, you are right - let's not

File: README.md [hunk 4]
```diff
@@ -11,13 +11,19 @@
 
 This package provides a core interface for working with [Markov decision processes (MDPs)](https://en.wikipedia.org/wiki/Markov_decision_process) and [partially observable Markov decision processes (POMDPs)](https://en.wikipedia.org/wiki/Partially_observable_Markov_decision_process). For examples, please see [POMDPExamples](https://github.com/JuliaPOMDP/POMDPExamples.jl), [QuickPOMDPs](https://github.com/JuliaPOMDP/QuickPOMDPs.jl), and the [Gallery](https://github.com/JuliaPOMDP/POMDPGallery.jl).
 
+---
+
+**NOTE**: We are currently in the process of upgrading solvers to work with POMDPs v0.9. This process is expected to take several weeks. For compatibility with all solvers, please install POMDPs v0.8 with `pkg"add POMDPs@v0.8"`.
```
[the remark this answers is not in the document]
Comment (zsunberg) [reply README.md:16]: It doesn't seem to work in github READMEs

File: README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (rejuvyesh) [reply README.md]: My bad. The only way would be to link to the markdown. But this is better than that.

**Pull request merged**